src/ripple/app/tx/impl/XChainBridge.cpp

```diff
@@ -0,0 +1,1675 @@
+        // cases (the attesters did their job)
+        STAmount const share = [&] {
+            STAmount const den{rewardAccounts.size()};
+            return divide(rewardPool, den, rewardPool.issue());
+        }();
+        STAmount distributed = rewardPool.zeroed();
+        for (auto const& ra : rewardAccounts)
+        {
+            auto const thTer = transferHelper(
+                psb,
+                rewardPoolSrc,
+                ra,
+                /*dstTag*/ std::nullopt,
+                // claim owner is not relevant to distributing rewards
+                /*claimOwner*/ std::nullopt,
+                share,
+                TransferHelperCanCreateDst::no,
+                j);
+
+            if (thTer == tecINSUFFICIENT_FUNDS || thTer == tecINTERNAL)
+                return thTer;
+
+            if (isTesSuccess(thTer))
+                distributed += share;
+
+            // let txn succeed if error distributing rewards (other than
+            // inability to pay)
+        }
+
+        if (distributed > rewardPool)
+            return tecINTERNAL;  // LCOV_EXCL_LINE
+    }
+
+    return thTer;
+}
+
+/** Get signers list corresponding to the account that owns the bridge
+
+    @param view View to read the signer's list from.
+    @param sleB Sle of the bridge.
+    @param j Log
+
+    @return map of the signer's list (AccountIDs and weights), the quorum, and
+            error code
+
+    @note If the account includes a regular key or master key, it is included
+   with the signer's list with a maximum weight. If the account does not include
+   a signer's list, the threshold is set to 1 (i.e. either the master key or
+   regular key can sign)
+*/
+std::tuple<std::unordered_map<AccountID, std::uint32_t>, std::uint32_t, TER>
+getSignersListAndQuorum(ApplyView& view, SLE const& sleB, beast::Journal j)
+{
+    std::unordered_map<AccountID, std::uint32_t> r;
+    std::uint32_t q = std::numeric_limits<std::uint32_t>::max();
+
+    AccountID const thisDoor = sleB[sfAccount];
+    auto const sleDoor = [&] { return view.read(keylet::account(thisDoor)); }();
+
+    if (!sleDoor)
+    {
+        return {r, q, tecINTERNAL};  // LCOV_EXCL_LINE
+    }
+
+    auto const masterKey = [&]() -> std::optional<AccountID> {
+        if (sleDoor->isFlag(lsfDisableMaster))
+            return std::nullopt;
+        return thisDoor;
+    }();
+
+    std::optional<AccountID> regularKey = (*sleDoor)[~sfRegularKey];
+
+    auto const sleS = view.read(keylet::signers(sleB[sfAccount]));
+    if (!sleS)
+    {
+        if (masterKey || regularKey)
+        {
+            q = 1;
+            if (masterKey)
+                r[*masterKey] = std::numeric_limits<std::uint16_t>::max();
+            if (regularKey)
+                r[*regularKey] = std::numeric_limits<std::uint16_t>::max();
+
+            return {std::move(r), q, tesSUCCESS};
+        }
+        return {r, q, tecXCHAIN_NO_SIGNERS_LIST};
+    }
```

Comment:
It looks like this code is intended to allow an attestor to sign an attestation using the Door account's master or RegularKey.  It that's what is happening I don't think that's a good idea.

The attestations are modeled after commonly used multisignatures in the XRPLedger.  You can't use the master key or regular key of the signed-for account if you use multisignatures.  You can _only_ use the multisigners in the `SignerList` owned by the account.

My strong recommendation would be to remove the code that allows a master or regular key to be used in this way.

I'm afraid that this change was the result of misunderstanding a comment I made earlier: https://github.com/XRPLF/rippled/pull/4292#issuecomment-1284495909

If that's the case, then this change entirely misses the intended mark.

I've made an example of what I intended to request in https://github.com/XRPLF/rippled/pull/4292#issuecomment-1284495909.  You can see that example in the top-most commit here: https://github.com/scottschurr/rippled/commits/scottd-xbridge-reg-key

Reply:
This is also a feature because I requested it - it's a lot easier to bootstrap a bridge when you don't need to create a temporary multisign in order to set up the initial accounts (e.g. the witness submission accounts). 

I believe the reason it was believed to be fine to do it this way is because the threat model is the same regardless - you need to trust whoever controls the account, whether that is via owning the master key/regular key or via a quorum of multisigners, since they could theoretically sign any transaction to mishandle those funds. Also, any properly set up bridge should disable the master key and should not have a regular key enabled.

Reply:
I think it's OK to keep this feature to make @mvadari life easier. It's easy to justify from a security point of view - the regular key and master key can move funds on their own from the door account, so they should be able to sign attestations on their own. Since it doesn't cause harm and makes writing her scripts easier I'd like to keep it. I understand where @scottschurr is coming from trying to make attestation the same as multisig. I think allowing master and regular keys makes it the same as signing a transaction from the account, which can be done with both a multisig list and a regular or master key.

We can discuss this further tho.

Reply:
I'll bring this up with @mvardari again. I think the doc would say "whatever you can use to sign a transaction on the door account can also be used to sign an attestation: master key (if enabled), regular key (if set), or signer's list)"

Reply:
I removed this in `f8284b0675 [fold] Do not allow the master or regular key to sign attestations` (attn: @scottschurr )